From 622c664c5ed10f3773460d503c2853b74b33ab5e Mon Sep 17 00:00:00 2001
From: Harsh Kumar Bijlani <hbijlani@qti.qualcomm.com>
Date: Tue, 1 Jul 2025 15:39:09 +0530
Subject: [PATCH] UPSTREAM: CP: wifi: ath12k: Use random mac address for scan
 vdev

Currently for scanning purpose, there is a separate vdev created and
this vdev uses the same mac address as that of the existing AP vdev.

Due to the same mac address usage, the ath12k_dp_peer of the scan vdev and
AP vdev point to the same ath12k_dp_peer and when this scan vdev is deleted,
then ath12k_dp_peer also gets deleted. This results in use-after-free of
ath12k_dp_peer for AP vdev in the Tx path.

In order to fix this, generate and make use of random mac address for
scan vdev.

Signed-off-by: Harsh Kumar Bijlani <hbijlani@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index cdd10fc..0c51ec2 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -13070,10 +13070,7 @@ int ath12k_mac_vdev_create(struct ath12k *ar, struct ath12k_link_vif *arvif,
 			}
 		}
 
-		if (arvif->link_id == ATH12K_DEFAULT_SCAN_LINK &&
-		    !is_zero_ether_addr(arvif->bssid)) {
-			memcpy(link_addr, arvif->bssid, ETH_ALEN);
-		} else if (link_conf) {
+		if (link_conf && !arvif->is_scan_vif) {
 			memcpy(link_addr, link_conf->addr, ETH_ALEN);
 			memcpy(arvif->bssid, link_conf->addr, ETH_ALEN);
 		} else {
-- 
2.34.1

